### PR TITLE
Add a text on driver enabling

### PR DESCRIPTION
**Pull request by ereshetova**
Signed-off-by: Elena Reshetova <elena.reshetova@intel.com>

### CODE REVIEW

**Comment by ereshetova** [comment]
@sirmc Could you please check that I didnt make any mistakes by embedding  (and slightly rewriting) your fuzzing text into the overall flow?

File: source/tdx-guest-hardening.rst [hunk 1]
```diff
@@ -1000,4 +1000,325 @@ https://github.com/intel/ccc-linux-guest-hardening/tree/master/bkc/kafl/userspac
...
+everyone benefits
+the joint hardened kernel, as well to get suggestions on the most
+appropriate way of
+fixing these issues. Also in order to verify that the issues have been
+addressed by
+respective patches, a new round of fuzzing needs to be performed to
+verify that the
+issues found in step 3 are not reproducible anymore.
+
+5. **Enable driver in the TDX filter**. When the driver code has been
```
Comment (sirmc) [inline source/tdx-guest-hardening.rst]: I think we need to mention the filter before explaining how to setup the fuzzing harness. 1) reference this in the text or 2) move this section up or 3) suggest disabling the filter all together

File: source/tdx-guest-hardening.rst [hunk 2]
```diff
@@ -1000,4 +1000,325 @@ https://github.com/intel/ccc-linux-guest-hardening/tree/master/bkc/kafl/userspac
...
+	modprobe vhost_vsock
+	chmod 0666 /dev/vhost-vsock
+	qemu: -device vhost-vsock-pci,id=vhost-vsock-pci0,guest-cid=3
+
+**Guest steps**. The next step in enabling virtio-vsock fuzzing is to
+set up the kAFL userspace fuzzing harness in the following way.
+
+First, the guest kernel needs to be compiled with vsock support
+(CONFIG_VIRTIO_VSOCKET=y and CONFIG_VHOST_VSOCK=y). Alternatively, it
+can be also enabled as a kernel module, but his will require an
```
Comment (Wenzel) [inline source/tdx-guest-hardening.rst]: typo: but his -> but this

File: source/tdx-guest-hardening.rst [hunk 3]
```diff
@@ -1000,4 +1000,325 @@ https://github.com/intel/ccc-linux-guest-hardening/tree/master/bkc/kafl/userspac
...
+**Example**. Enabling fuzzing targets like the virtio-vsock driver
+requires some manual work and modifications of the fuzzing setup (as
+opposite to more straightforward examples like virtio-net or
+virtio-console) and below steps explain how to add support for such a
+target. In a nutshell, virtio-vsock sets up a socket on the host/VMM,
+allowing a host process to setup a direct socket connection to the
+guest VM over VirtIO. For fuzzing, this requires some initial setup in
+the host, as well as establishing a connection from the guest.
+
+**Host steps**. First, the VMM host kernel must support VSOCK. The
```
Comment (Wenzel) [inline source/tdx-guest-hardening.rst]: This paragraph is really difficult to follow. I think it should be broken down into multiple steps.

File: source/tdx-guest-hardening.rst [hunk 4]
```diff
@@ -1000,4 +1000,325 @@ https://github.com/intel/ccc-linux-guest-hardening/tree/master/bkc/kafl/userspac
...
+hardened and all
+the patches are integrated and verified, the driver can be enabled in
+the TDX guest by modifying
+the allow list in the TDX driver filter code in arch/x86/kernel/tdx-filter.c.
+
+**Example**. For the virtio-vsock driver the following patch adds it
+to the list of
+allowed devices on the virtio bus.
+
+   .. code-block:: c
```
Comment (Wenzel) [inline source/tdx-guest-hardening.rst]: This results in
~~~
tdx-guest-hardening.rst:1289: WARNING: Could not lex literal_block as "c". Highlighting skipped.
~~~

I propose to use `diff` as the langage here.

File: source/tdx-guest-hardening.rst [hunk 5]
```diff
@@ -1000,4 +1000,325 @@ https://github.com/intel/ccc-linux-guest-hardening/tree/master/bkc/kafl/userspac
...
+Intel TDX <https://github.com/intel/tdx-tools>`_ only enables a small set of
+virtio drivers that are essential for the TDX guest basic functionality. These
+drivers have been hardened using the methodology described in this document,
+but naturally different deployment scenarios and use cases for the TDX will
+require many more additional drivers to be enabled in the TDX guest kernel.
+
+This section provides guidance on how to use the methodology presented
+in this document for adding and hardening a new driver for the TDX guest kernel.
+
+Step-by-step guidance
```
Comment (Wenzel) [inline source/tdx-guest-hardening.rst]: You are declaring only one section here, and no other section in the rest of this chapter.
This ends up with a strange hierarchy:
image

Maybe you could instead use each step as a separate section ?

File: source/tdx-guest-hardening.rst [hunk 6]
```diff
@@ -1000,4 +1000,325 @@ https://github.com/intel/ccc-linux-guest-hardening/tree/master/bkc/kafl/userspac
...
+
+   .. code-block:: bash
+
+      ~/smatch_scripts/kchecker net/vmw_vsock/virtio_transport.c > driver_results
+
+The driver_results output file will contain the list of input points
+and the limited
+propagation information:
+
+   .. code-block:: shell
```
Comment (Wenzel) [inline source/tdx-guest-hardening.rst]: The `.. code-block:: shell` directive shouldn't be indented.
That's why the parsing failed here and the output looses the line breaks:
image


Instead, by removing the indentation like I suggest:
image

Results in:
image

**Review by Wenzel — COMMENTED** [review]
This is a first review on formatting, see my comments.
A general suggestion I would make is to try to make use of inline code highlighting, since it helps to clarify what is an acronym, a technical term, a command line from what is just english:

Example:
First, the VMM host kernel must support VSOCK. The corresponding kernel module can be loaded using modprobe vhost_vsock. If this fails, it might be required to install a different kernel which has CONFIG_VHOST_VSOCK set. When the vhost_vsock driver is enabled, a device shall appear at /dev/vhost-vsock. Its default permissions might be insufficient for QEMU to access, but it can be fixed by executing chmod 0666 /dev/vhost-vsock.

⬇️ 
First, the _VMM_ host kernel must support _VSOCK_. The corresponding kernel module can be loaded using `modprobe vhost_vsock`. If this fails, it might be required to install a different kernel which has `CONFIG_VHOST_VSOCK` set. When the `vhost_vsock` driver is enabled, a device shall appear at `/dev/vhost-vsock`. Its default permissions might be insufficient for _QEMU_ to access, but it can be fixed by executing `chmod 0666 /dev/vhost-vsock`.

File: source/tdx-guest-hardening.rst [hunk 7]
```diff
@@ -1000,4 +1000,325 @@ https://github.com/intel/ccc-linux-guest-hardening/tree/master/bkc/kafl/userspac
...
+/dev/vhost-vsock`. Now that the vhost-vsock device is available to
+QEMU, the device for the guest VM can be enabled by appending the
+string `-device vhost-vsock-pci,id=vhost-vsock-pci0,guest-cid=3` to
+QEMU options. The guest-cid value is a connection identifier that
+needs to be unique for the system. In other words, when fuzzing with
+multiple workers, each QEMU instance must use a separate guest-cid.
+For kAFL we have added some syntax magic to allow for these
+kinds of situations. In your `kafl_config.yaml` (by default found in
+`$BKC_ROOT/bkc/kafl/kafl_config.yaml`),  the following string can be
+appended to the `qemu_base` entry: `-device vhost-vsock-pci,id=vhost-vsock-pci0,guest-cid={QEMU_ID + 3}`.
```
Comment (sirmc) [inline source/tdx-guest-hardening.rst]: I haven't implemented the {QEMU_ID + 3} thing yet. Will try to push this out this week.

**Comment by ereshetova** [comment]
This update should address everything with exception of Mathieu's comment on the host fuzzing steps: "This paragraph is really difficult to follow. I think it should be broken down into multiple steps.". I am not sure how to break things logically here ((

**Comment by Wenzel** [comment]
I added links, and improved the formatting using interpreted text (with backticks) and inline code blocks where necessary.
- interpreted text: `` `QEMU` ``
- inline code blocks: `` :code:`virtio-vsock` ``

**Comment by ereshetova** [comment]
@Wenzel thank you very much! Let me merge this that it does not hang here while I am way for almost 3 weeks!

**Pull request merged**